engine/lib/cache.php

```diff
@@ -308,13 +308,67 @@ function _elgg_cache_init() {
 	_elgg_services()->systemCache->init();
 }
 
+/**
+ * Disable all caches
+ *
+ * @return void
+ * @internal
+ * @access private
+ */
+function _elgg_disable_caches() {
+	_elgg_services()->boot->getCache()->disable();
+	_elgg_services()->plugins->getCache()->disable();
+	_elgg_services()->sessionCache->disable();
+	_elgg_services()->dataCache->disable();
+	_elgg_services()->dic_cache->getCache()->disable();
+	_elgg_services()->simpleCache->disable();
+	_elgg_services()->systemCache->disable();
+	_elgg_services()->autoloadManager->getCache()->disable();
+}
+
+/**
+ * Clear all caches
+ *
+ * @return void
+ * @internal
+ * @access private
+ */
+function _elgg_clear_caches() {
+	_elgg_services()->boot->invalidateCache();
+	_elgg_services()->plugins->clear();
+	_elgg_services()->sessionCache->clear();
+	_elgg_services()->dataCache->clear();
+	_elgg_services()->dic_cache->flushAll();
+	_elgg_services()->simpleCache->invalidate();
+	_elgg_services()->fileCache->clear();
+	_elgg_services()->autoloadManager->deleteCache();
+}
+
+/**
+ * Enable all caches
+ *
+ * @return void
+ * @internal
+ * @access private
+ */
+function _elgg_enable_caches() {
+	_elgg_services()->boot->getCache()->enable();
+	_elgg_services()->plugins->getCache()->enable();
+	_elgg_services()->sessionCache->enable();
+	_elgg_services()->dataCache->enable();
+	_elgg_services()->dic_cache->getCache()->enable();
+	_elgg_services()->simpleCache->enable();
+	_elgg_services()->systemCache->enable();
+	_elgg_services()->autoloadManager->getCache()->enable();
+}
+
 /**
  * @see \Elgg\Application::loadCore Do not do work here. Just register for events.
  */
 return function(\Elgg\EventsService $events, \Elgg\HooksRegistrationService $hooks) {
 	$events->registerHandler('ready', 'system', '_elgg_cache_init');
-	
-	// register plugin hooks for cache reset
-	$events->registerHandler('cache:flush', 'system', 'elgg_reset_system_cache');
-	$events->registerHandler('cache:flush', 'system', 'elgg_invalidate_simplecache');
+
+	$events->registerHandler('cache:flush:before', 'system', '_elgg_disable_caches');
+	$events->registerHandler('cache:flush', 'system', '_elgg_clear_caches');
+	$events->registerHandler('cache:flush:after', 'system', '_elgg_enable_caches');
```

Comment:
couldn't this cause potential problems? If (for some reason) I disabled a cache, then flush (all) the caches it gets enabled again. Not what i expected.

Reply:
Just use the handlers directly then.

Reply:
All caches are private API anyway, plugins have no business messing with them. System cache enabled/disabled is handled separately I think. We can just inject a black hole instead of a file cache if need be.